draft-chen-cfrg-vdaf-pine.md

```diff
@@ -346,10 +347,80 @@ implements this encoding step.
 
 ## The FLP Circuit
 
-> XXX Give an overview of the circuit's operation. Here it would be appropriate
-> to name all of the intermediate results: the bit checks, the wraparound
-> checks result (the reduction over all of the checks), the wraparound success
-> count, the squred norm equality test, and the squared norm range check.
+The validity circuit unpacks the encoded measurement as the following:
+
+* The first `dimension` entries are the `encoded_gradient`, the field elements
+  encoded from the floating point numbers.
+* The next `bit_checked_len` entries are expected to be bits, and should contain
+  the bits from L2-norm check, the bits from the range check in wraparound
+  checks, and the success bits in wraparound checks.
+* The last `num_wr_checks` are the wraparound check results, i.e., the dot
+  products of the encoded gradient and the random vectors.
+
+It also unpacks the joint randomness to compute random linear combinations of
+all the checks:
+
+* The first joint randomness field element is to reduce over the bit checks at
+  all bit entries.
+* The second joint randomness field element is to reduce over all the quadratic
+  checks in wraparound check.
+* The last joint randomness field element is to reduce over all the checks,
+  which include the reduced bit check result, the L2-norm equality check, the
+  L2-norm range check, the reduced quadratic checks in wraparound check, and
+  the success count check for wraparound check.
+
+Some of the auxiliary functions are defined in {{pine-auxiliary}}.
+
+### Bit Check {#valid-bit-check}
+
+The bit check computes a random linear combination of the polynomial `x(x-1)`
+evaluated at all bit entries of `bit_checked`, with the coefficients computed
+from the joint randomness `r_bit_check`. Note we need to normalize the constant
+`1` in the polynomial, by multiplying it by the inverse of number of shares
+`shares_inv`.
+
+> TODO Put `PineValid.eval_bit_check()` implementation here.
+
+### L2-Norm Check {#valid-norm-check}
+
+The validity circuit verifies two properties for L2-norm check:
+
+* Equality check: The computed sum of squares of `encoded_gradient` is equal to
+  the squared L2-norm decoded from the bit representation `sq_norm_v_bits`
```

Comment:
I don't think variable `sq_norm_v_bits` has not been defined? Ask yourself if you actually need to define it in order to convey what's going on.

I think we should try to explain what's going on without referring to variables in the circuit. What would help is an explanation of how range checking works in PINE. That way all you need to do here is say that the circuit decodes the bit-encoding of the reported norm and performs the range check.

Likewise for "wraparound check" section below. the explanation of how range checking works should go somewhere towards the top of the "FLP Circuit" section.

Reply:
The explanation is great. I think we can now simplify this text accordingly, by not mentioning variables. Also, I suggest describing what the circuit actuall does here:

> Equality check: The squared norm of the gradient is equal to the value reported by the Client. For this, the Aggregators decode their shares of the bit-representation of the squared norm (as defined above in {{valid-bit-check}}), compute their shares of the squared norm from their shares of the gradient, and check that the values are equal.
> Range check: The squared norm reported by the Client is in the desired range. For this, the Aggregators run the range check described in {{range-check}}.

Likewise for the other "sub-checks".